drake/systems/frames/LCMCoordinateFrame.cpp

```diff
@@ -0,0 +1,125 @@
+#include "LCMCoordinateFrame.h"
+#include <thread>
+#include "timeUtil.h"
+
+using namespace std;
+using namespace Eigen;
+
+bool encode(const CoordinateFrame* frame, const double t, const VectorXd& x, drake::lcmt_drake_signal& msg) {
+  msg.timestamp = static_cast<int64_t>(t*1000);
+  msg.dim = frame->getDim();
+  msg.coord = frame->getCoordinateNames(); // note: inefficient to do a deep copy every time
+  for (int i=0; i<msg.dim; i++) msg.val.push_back(x(i));
+  return true;
+}
+
+bool decode(const CoordinateFrame* frame, const drake::lcmt_drake_signal& msg, double& t, VectorXd& x) {
+  t = double(msg.timestamp)/1000.0;
+  for (int i=0; i<msg.dim; i++) { x(i) = msg.val[i]; }  // todo: make this more robust by doing string matching on the coordinate names (with a hashmap?)
+  return true;
+}
+
+
+bool waitForLCM(lcm::LCM* lcm, double timeout) {
+  int lcmFd = lcm->getFileno();
+
+  struct timeval tv;
+  tv.tv_sec = 0;
+  tv.tv_usec = timeout * 1e6;
+
+  fd_set fds;
+  FD_ZERO(&fds);
+  FD_SET(lcmFd, &fds);
+
+  int status = select(lcmFd + 1, &fds, 0, 0, &tv);
+  if (status == -1 && errno != EINTR)
+  {
+    cout << "select() returned error: " << errno << endl;
+  }
+  else if (status == -1 && errno == EINTR)
+  {
+    cout << "select() interrupted" << endl;
+  }
+  return (status > 0 && FD_ISSET(lcmFd, &fds));
+}
+
+class LCMLoop {
+public:
+  bool stop;
+  lcm::LCM* lcm;
+
+  LCMLoop(const shared_ptr<lcm::LCM>& _lcm) : lcm(_lcm.get()), stop(false) {}
```

Comment:
again, i don’t really think so.  the lifetime of the lcmloop object is inside a single function call of the owner of that pointer.  it would be wasteful to force increment counts on a shared pointer, etc.

> On Aug 22, 2015, at 9:49 PM, Twan Koolen <notifications@github.com> wrote:
> 
> In drake/systems/frames/LCMCoordinateFrame.cpp <https://github.com/RobotLocomotion/drake/pull/1274#discussion_r37701775>:
> 
> > +  {
> > +    cout << "select() returned error: " << errno << endl;
> > +  }
> > +  else if (status == -1 && errno == EINTR)
> > +  {
> > +    cout << "select() interrupted" << endl;
> > +  }
> > +  return (status > 0 && FD_ISSET(lcmFd, &fds));
> > +}
> > +
> > +class LCMLoop {
> > +public:
> > +  bool stop;
> > +  lcm::LCM* lcm;
> > +
> > +  LCMLoop(const shared_ptr<lcm::LCM>& _lcm) : lcm(_lcm.get()), stop(false) {}
> Again, very dangerous to call .get() on a shared_ptr and store the raw pointer.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/RobotLocomotion/drake/pull/1274/files#r37701775>.
>